docs(adr): ADR-21: Subspace specific custom fee tokens (#1119)
desmos-labs/desmos: docs/architecture/adr-021-subspace-specific-custom-fee-tokens.md

# ADR 021: Subspace specific custom fee tokens

## Changelog

- April 24th, 2023: First draft;
- May 8th, 2023: First review;

## Status

> Proposed

Reviewer: ```suggestion
ACCEPTED Not Implemented
```

Ideally, the flow that will lead one subspace to accept an additional fee token is going to be the following: 

1. The subspace owner creates an on-chain governance proposal asking validators if they are fine in receiving fees paid in the new token denomination when validating transactions related to that subspace.
2. Validators will agree or reject the proposal through on-chain voting. 
3. If the proposal is accepted, the new token denom will be added to the list of additional fee token denoms that can be used to pay for fees within that subspace.
4. If the proposal is rejected, the new token denom will not be added to the list of additional fee token denoms.

### Types

To represent the list of supported fee token denoms and their default minimum prices within a subspace, we will add a new field `allowed_fee_tokens` to the existing Subspace structure.

```proto
message Subspace {

  ...skip

  // the creation time of the subspace
  google.protobuf.Timestamp creation_time = 7;
  
  // List of fee token denoms with default minimum gas prices allowed inside the subspace
  repeated Coin allowed_fee_tokens = 8;
}
```

### Msgs

> We will implement the operation that allows subspace admins to updates subspace allowed fee tokens list by governance;

Reviewer: ```suggestion
We will implement the operation that allows subspace admins to update the list of allowed fee tokens list within their subspace via a governance proposal:
```

```proto
service Msg {

>   // UpdateSubspaceFeeTokens allows subspace admins to update the allowed tokens to be fee tokens inside the subspace by governance

Reviewer: ```suggestion
  // UpdateSubspaceFeeTokens allows subspace admins to update the list of fee tokens inside the subspace via a  governance proposal
```

}

// MsgUpdateSubspaceFeeTokens represents the message to be used to update a subspace fee tokens by governance
message MsgUpdateSubspaceFeeTokens {

  // Id of the subspace where the list of allowed fee tokens will be updated
  uint64 subspace_id = 1;
    
  // List of the allowed tokens to be fee token inside the subspace along with their default minimum prices
  repeated Coin allowed_fee_tokens = 2;
    
  // authority is the address that controls the module (defaults to x/gov unless overwritten).
  string authority = 3;
}

// MsgUpdateSubspaceFeeTokensResponse represents the Msg/UpdateSubspaceFeeTokens response type
message MsgUpdateSubspaceFeeTokensResponse {}
```

### Custom TxFeeChecker

To make it easier for validators to manage the minimum prices of fee tokens allowed within a subspace, we will develop a new `TxFeeChecker` based on the existing one, which will function as follows:
- Combine the list of minimum gas prices in the validator's local configuration with the list of allowed fee tokens and their minimum prices within the subspace.
- Follow the same process as the existing `TxFeeChecker`.

### Deprecated `x/fees`

> Since `x/fees` module may conflict with the custom fee tokens feature, as demonstrated in the following scenario:
> - Governance decides to change the minimum fees of MsgCreatePost to `10dsm` using `x/fees`.
> - Governance accepts the custom token `minttoken` as a fee within subspace 1 using `MsgUpdateSubspaceFeeTokens`.
> 
> Due to this conflict, `minttoken` will not be accepted as a fee within the subspace 1. Therefore, we decided to drop `x/fees` from Desmos.

Reviewer: ```suggestion
By implementing the mechanism described above, there might be a problem with the `x/fees` module. Here's an example scenario to illustrate this:

1. the governance decides to use `x/fees` to change the minimum fees of `MsgCreatePost` to `10dsm`.
2. the governance also wants to use a custom token called `TokenA` as a fee token within subspace `1` using `MsgUpdateSubspaceFeeTokens`.

Because `x/fees` takes priority over anything, `TokenA` will not be accepted as a fee token within subspace `1`. 

For this reason, we have decided to remove the `x/fees` module from Desmos.
```

## Consequences

### Backwards Compatibility

> The solution outlined above is **not** backwards compatible and will require a migration script to update all existing subspaces to the new version. This script will handle the following tasks:
> - migrate all subspaces to have a default allowed fee tokens list.

Reviewer: ```suggestion
The solution outlined above is **not** backward compatible, and it requires a migration script to update all existing subspaces to the new version. The script should take care of migrating all subspaces to have a default allowed fee tokens list.
```

Reviewer: Is this actually true though? If we mark the Proto field as [`optional`](https://protobuf.dev/programming-guides/proto3/#specifying-rules), do we need to migrate everything?

Author: Now proto3 fields are all optional now. You are right, adding new field does not cause the backward compatibility issues. I just learned from the official documentation. Thanks!

> https://protobuf.dev/programming-guides/proto3/#updating
If you add new fields, any messages serialized by code using your “old” message format can still be parsed by your new generated code. You should keep in mind the [default values](https://protobuf.dev/programming-guides/proto3/#default) for these elements so that new code can properly interact with messages generated by old code. Similarly, messages created by your new code can be parsed by your old code: old binaries simply ignore the new field when parsing. See the [Unknown Fields](https://protobuf.dev/programming-guides/proto3/#unknowns) section for details.

```proto
service Msg {
  // UpdateSubspaceFeeTokens allows subspace admins to update the list of fee tokens inside the subspace via a governance proposal
  rpc UpdateSubspaceFeeTokens(MsgUpdateSubspaceFeeTokens) returns (MsgUpdateSubspaceFeeTokensResponse);
}

// MsgUpdateSubspaceFeeTokens represents the message to be used to update a subspace fee tokens by governance
message MsgUpdateSubspaceFeeTokens {

  // Id of the subspace where the list of allowed fee tokens will be updated
  uint64 subspace_id = 1;
    
  // List of the allowed tokens to be fee token inside the subspace along with their default minimum prices
  repeated Coin allowed_fee_tokens = 2;
    
  // authority is the address that controls the module (defaults to x/gov unless overwritten).
  string authority = 3;
}

// MsgUpdateSubspaceFeeTokensResponse represents the Msg/UpdateSubspaceFeeTokens response type
message MsgUpdateSubspaceFeeTokensResponse {}
```

### Custom TxFeeChecker

To make it easier for validators to manage the minimum prices of fee tokens allowed within a subspace, we will develop a new `TxFeeChecker` based on the existing one, which will function as follows:

> - Combine the list of minimum gas prices in the validator's local configuration with the list of allowed fee tokens and their minimum prices within the subspace.
> - Follow the same process as the existing `TxFeeChecker`.

Reviewer: ```suggestion
1. combine the list of minimum gas prices in the validator's local configuration with the list of allowed fee tokens and their minimum prices within the subspace;
2. follow the same process used inside the existing `TxFeeChecker`.
```

# ADR 021: Subspace specific custom fee tokens

## Changelog

- April 24th, 2023: First draft;

## Status

Proposed

## Abstract

> This ADR introduces a mechanism that allows users to pay fees within a designated subspace using their existing service tokens.

Reviewer: ```suggestion
This ADR enables subspace owners to let content creators pay for transaction fees using different token denominations when creating contents within their subspace.
```

## Context

> Desmos uses subspaces to represent virtual applications built on top of our protocol. Currently, subspaces are mainly used to store content such as posts, reports, and reactions. While this is sufficient for developing DApps on Desmos, it may not be enough to convince existing applications to migrate to our platform since Desmos is lack of a mechanism that enables users to pay fees with existing service tokens.
> 
> For instance, consider Keplr's on-chain governance discussions feature, which is currently implemented using a custom backend that stores data in a centralized way. Although Keplr could easily use Desmos as their backend, they are not willing to do so because it offers no advantage to them. Migrating to Desmos would simply transfer their existing users to our chain without any benefits for them.

Reviewer: ```suggestion
Desmos uses subspaces to represent virtual applications built on top of our protocol. Currently, subspaces are mainly used to store content such as posts, reports, and reactions. While these features are sufficient to develop new applications on top of Desmos, they may not be enough to convince existing applications to migrate to our platform.

For instance, let's consider an exiting application that uses a centralized database in order to store users discussions. Although such application might easily use Desmos as its backend, they most likely are not going to do so because it offers no advantage to them. Migrating to Desmos would simply transfer their existing users to our chain without any benefits for them.
```

## Decision

> To address the issue mentioned above, we propose implementing a mechanism that allows users to pay fees using the existing service tokens within the specified subspace. The mechanism must have the following features:
> - Defining a minimum gas prices for non-DSM tokens that validators are willing to accept via governance proposal;
> - Allowing the subspace manager(s) to update the list of fee tokens allowed in the subspace;
> - Checking the on-chain minimum gas price requirement for non-DSM fees using a custom `TxChecker`.

Reviewer: ```suggestion
To address the issue mentioned above, we propose implementing a mechanism that allows users to pay content-related fees using additional token denominations within the specified subspace. 

Ideally, the flow that will lead one subspace to accept an additional fee token is going to be the following: 

1. The subspace owner creates an on-chain governance proposal asking validators if they are fine in receiving fees paid in the new token denomination when validating transactions related to that subspace.
2. Validators will agree or reject the proposal through on-chain voting. 
3. If the proposal is accepted, the new token denom will be added to the list of additional fee token denoms that can be used to pay for fees within that subspace.
4. If the proposal is rejected, the new token denom will not be added to the list of additional fee token denoms.
```

> The upcoming changes are as follows:

Reviewer: ```suggestion
```

### Types

We will define a new `Params` type to show the `x/subspaces` parameters.

```proto
// Params contains the module parameters
message Params {
  // List of the minimum gas prices to be accepted by validators
  repeated Coin min_gas_prices = 2;
}

> ```

Reviewer: Why is this needed? Each subspace will have its own `allowed_fee_tokens` list anyway. There's no need to have a module-wide list

> In addition, we will add a __allowed_fee_tokens__ field to current `Subspace` structure.

Reviewer: ```suggestion
In order to represent the list of fee token denoms that are supported by a subspace, we will add a new field to the current `Subspace` structure.
```

```proto
message Subspace {

>   // Unique id that identifies the subspace
>   uint64 id = 1;
> 
>   // Human-readable name of the subspace
>   string name = 2 [ (gogoproto.moretags) = "yaml:\"name\"" ];

Reviewer: No need to put theses. Thye can be included inside the `...skip` part

```suggestion
```

  ...skip

  // the creation time of the subspace
  google.protobuf.Timestamp creation_time = 7;
  

>   // List of allowed fee tokens inside the subspace

Reviewer: ```suggestion
  // List of fee token denoms allowed inside the subspace
```

  repeated string allowed_fee_tokens = 8;
}
```

### Msgs

We will allow the the following operations:
1. update subspace allowed fee tokens list by managers;

> 2. update `x/subspaces` parameters by governance.

Reviewer: Why is this needed? It would be much simpler to require subspace owners to go through a governance-gated update of the subspace `allowed_fee_tokens` list instead of having two different message types for this. Just: 

1. The subspace owner creates an on-chain proposal to update the list of tokens 
2. The governance votes yes/no 
3. Based on the result, the list is updated or not.

No need for multiple message types or anything like that

Author: Splitting the message into two parts would make it easier for validators to manage min-gas-prices. 

For example, let's say Subspace A and B both allow a min-gas-price of 0.1 minttoken. If validators want to update the mini-gas-price for minttoken to 0.2 minttoken after a market crash, they would need to perform the following operations:
1. Propose a new min-gas-price of 0.2 minttoken for Subspace A
2. Propose a new min-gas-price of 0.2 minttoken for Subspace B
3. Propose multiple proposals to update the min-gas-price for all subspaces that allow minttoken

However, this approach can make it difficult for validators to manage min-gas-prices. To address this, I decided to create a separate list for minimum gas prices.

Reviewer: @dadamu This is extremely complicated to understand and wrap my head around. I honestly cannot understand: 

1. why there's the need to have two different messages.  
   If both subspace managers *and* any other user must use *the same* governance proposal to update the subspace fee tokens, why do we need to have two different messages? 
2. why can't we use an implementation of the `Content` interface, rather than custom messages?  
   Why can't we have something as simple as other governance proposal types such as `SpendCommunityProposal` but instead we need to have this custom message? 

I honestly don't think it's necessary to have this extremely complicated mechanism with two different messages types. Instead, we can simply use a single `Content` implementation type and rely on that

Author: @RiccardoM `Content` interface is gov v1beta1 feature, it will be deprecated in the future. Now, cosmos-sdk applies proposal message to have spend community proposal. See:
https://github.com/cosmos/cosmos-sdk/blob/v0.47.2/proto/cosmos/distribution/v1beta1/tx.proto#L155

Reviewer: @dadamu Ok, I see what you mean about the `Content` interface. 

For the other discussion, about two messages, it makes little to no sense to have such an incredible complex procedure. Instead, the procedure should be as follow: 

1. The subspace admin creates a governance proposal asking validators if it's OK for them to accept fees paid in the `TokenA` inside their subspace.
2. If the proposal passes, the list of fee tokens **only within that subspace** is updated. 

This guarantees that: 

1. Subspace admins **always** have to ask the governance its permission to accept a new token 
2. The governance can agree to accept `TokenA` as a fee token inside one subspace, but reject it for another subspace (for whatever reason)
3. There is a lower chance of introducing bugs: with a simplified procedure and less operations, fewer things can go wrong

To do this, we can simply implement a `MsgUpdateSubspaceFeeTokens` that accepts: 
- a `SubspaceID`
- a list of coin denominations 

It should be executed only by the governance module, making it de-facto as much similar as the old `Content` interface.

Author: @RiccardoM I see what you want to do is reduce the potential bugs, while this procedure is lack of the minimum gas prices management from validators since only Subspace `admin` can create a proposal to modify the minimum gas prices of subspaces. I believe validators would like to fix the minimum gas prices after market crash.

For example, multiple subspaces passed proposals to use stable coin `USDM`. `USDM` depegs and become 0.001USD after market crash. Then validator must ask all the subspaces managers to modify their own fee token list to remove `USDM`, or validators will keep receiving valueless `USDM` as fees.

Can you give me a sample that as a validator how to modify the minimum prices of `USDM` inside each subspace in a one proposal?

Reviewer: @dadamu If validators want to modify the minimum gas prices they accept, they should be able to do it by editing the `min-gas-price` inside their node configuration. We could simply read that file and if the value written there is less than the value inside the subspace, we should make sure the transaction is not added to the validator's mem pool. 

This way: 
- subspace admins are the only ones that can touch the fee denoms accepted within their subspace;
- if validators want to independently update their min gas prices, they can do that without a problem;
- if no validator sets a min-gas-price for a specific fee denom, then the default one for that subspace will be used.

Do you think that's technically possible?

Author: Yes, it is doable and solves my concerns. I will update ADR to include everything we discussed.

Author: @RiccardoM Updated.

}

// MsgUpdateSubspaceFeeTokens represents the message to be used to update a subspace fee tokens by governance
message MsgUpdateSubspaceFeeTokens {

  // Id of the subspace where the list of allowed fee tokens will be updated
  uint64 subspace_id = 1;
    
  // List of the allowed tokens to be fee token inside the subspace along with their default minimum prices
  repeated Coin allowed_fee_tokens = 2;
    
  // authority is the address that controls the module (defaults to x/gov unless overwritten).
  string authority = 3;
}

// MsgUpdateSubspaceFeeTokensResponse represents the Msg/UpdateSubspaceFeeTokens response type
message MsgUpdateSubspaceFeeTokensResponse {}
```

### Custom TxFeeChecker

To make it easier for validators to manage the minimum prices of fee tokens allowed within a subspace, we will develop a new `TxFeeChecker` based on the existing one, which will function as follows:
- Combine the list of minimum gas prices in the validator's local configuration with the list of allowed fee tokens and their minimum prices within the subspace.
- Follow the same process as the existing `TxFeeChecker`.

> ### Update `x/fees`

Author: Since `x/fees` may conflict with subspace fee token checks, I think we can update the logic of `CheckFees`. What do you think? @RiccardoM

> ### Update `x/fees`

Author: Since x/fees may conflict with subspace fee token checks, I think we can update the logic of CheckFees. What do you think? @RiccardoM @manu0466

Reviewer: I think we should completely drop the `x/fees` module instead. It's never been used anyway

Author: @RiccardoM Updated docs.
